### ISSUE TITLE
[Feature] Multi-components workload scheduling

### ISSUE THREAD

**Issue by RainbowMango** [karmada-io/karmada#5115]
**What would you like to be added**:
Provide fine-grained support for workloads which usually consist of two or more components, like:

- Kubeflow Training/Big-data workloads
  * [TFJob](https://github.com/kubeflow/training-operator/blob/2b39d3cbc1525dc29eb40aa58b149dbeef00aa0f/pkg/apis/kubeflow.org/v1/tensorflow_types.go#L50)(Tensorflow Job), which may consist of multiple components like `PS`, `Worker`, `Chief`, `Master`, and `Evaluator`.
  * [PyTorchJob](https://github.com/kubeflow/training-operator/blob/2b39d3cbc1525dc29eb40aa58b149dbeef00aa0f/pkg/apis/kubeflow.org/v1/pytorch_types.go#L56) which may consist of two components, like `Master` and `Worker`.
  * [MXJob](https://github.com/kubeflow/training-operator/blob/2b39d3cbc1525dc29eb40aa58b149dbeef00aa0f/pkg/apis/kubeflow.org/v1/mxnet_types.go#L112), which may consist of multiple components, like `Scheduler`, `Server`, `Worker`, `TunerTracker`, `TunerServer`, and `Tuner`.
  * [XGBoostJob](https://github.com/kubeflow/training-operator/blob/2b39d3cbc1525dc29eb40aa58b149dbeef00aa0f/pkg/apis/kubeflow.org/v1/xgboost_types.go#L66), which may consist of two components, like `Master` and `Worker`.
  * [MPIJob](https://github.com/kubeflow/training-operator/blob/2b39d3cbc1525dc29eb40aa58b149dbeef00aa0f/pkg/apis/kubeflow.org/v1/mpi_types.go#L51C6-L51C13), which may consist of two components, like `Launcher` and `Worker`.
  * [PaddleJob](https://github.com/kubeflow/training-operator/blob/2b39d3cbc1525dc29eb40aa58b149dbeef00aa0f/pkg/apis/kubeflow.org/v1/paddlepaddle_types.go#L56), which may consist of two components, like `Master` and `Worker`.
  * [SparkApplication](https://github.com/kubeflow/spark-operator/blob/16cd35a0a220da0304383517c73a5fdb36f4213c/pkg/apis/sparkoperator.k8s.io/v1beta2/types.go#L172), which may consist of two components, like `driver` and `executor`.
- [FlinkDeployment](https://nightlies.apache.org/flink/flink-kubernetes-operator-docs-release-1.9/), which may consist of two components, like `jobManager` and `taskManager`.
- [Vocalno Job](https://volcano.sh/en/docs/vcjob/), which may consist of multiple tasks.
- etc

**User story**

**Why is this needed**:

**Comment by Monokaix** [comment]
Replica division is a key issue for multi pod template workload, is there any resolution here?

**Comment by KunWuLuan** [comment]
Some other job types:

[RayJob](https://github.com/ray-project/kuberay/blob/b81af7c6e01433d51ded5a17edf6bb5a1dc76f88/ray-operator/apis/ray/v1/rayjob_types.go#L71)

[TrainJob](https://github.com/kubeflow/training-operator/blob/95be3c04967e6dc5a8def067d10113ecf78a2ef1/pkg/apis/kubeflow.org/v2alpha1/trainjob_types.go#L38)

**Comment by RainbowMango** [comment]
Thanks for the report. I will look at them and get back to you.

**Comment by RainbowMango** [comment]
For the RayJob, I can see the workers are declared with the [WokerGroupSpecs](https://github.com/ray-project/kuberay/blob/b5f14f1a173d32e6e5c5abfecd11d32a4ed780c1/ray-operator/apis/ray/v1/raycluster_types.go#L29), is there a `master`? (I don't know if you usually say master.)

**Comment by RainbowMango** [comment]
And For the `TrainJob`, I can see only one [Trainner](https://github.com/kubeflow/training-operator/blob/94dee0e19f0becc115469edeaefd0a1dedf3075e/pkg/apis/kubeflow.org/v2alpha1/trainjob_types.go#L112) for each TrainJob, so it seems not a multi-compoments workload.  Please correct me if I'm wrong. @KunWuLuan

**Comment by KunWuLuan** [comment]
> For the RayJob, I can see the workers are declared with the [WokerGroupSpecs](https://github.com/ray-project/kuberay/blob/b5f14f1a173d32e6e5c5abfecd11d32a4ed780c1/ray-operator/apis/ray/v1/raycluster_types.go#L29), is there a `master`? (I don't know if you usually say master.)

For a RayJob, there will be some [head pods](https://github.com/ray-project/kuberay/blob/b5f14f1a173d32e6e5c5abfecd11d32a4ed780c1/ray-operator/apis/ray/v1/raycluster_types.go#L25) and some worker groups. Worker pods communicate with head pods to run tasks.

**Comment by KunWuLuan** [comment]
> And For the `TrainJob`, I can see only one [Trainner](https://github.com/kubeflow/training-operator/blob/94dee0e19f0becc115469edeaefd0a1dedf3075e/pkg/apis/kubeflow.org/v2alpha1/trainjob_types.go#L112) for each TrainJob, so it seems not a multi-compoments workload. Please correct me if I'm wrong. @KunWuLuan

TrainJob will refer to a TrainingRuntime, which will refer to a [JobSet](https://jobset.sigs.k8s.io/docs/concepts/).

**Comment by RainbowMango** [comment]
> For a RayJob, there will be some [head pods](https://github.com/ray-project/kuberay/blob/b5f14f1a173d32e6e5c5abfecd11d32a4ed780c1/ray-operator/apis/ray/v1/raycluster_types.go#L25) and some worker groups. Worker pods communicate with head pods to run tasks.

Got it, thanks. Given worker group could have a different pod template, so each worker group can be treated as a component. Right?

**Comment by KunWuLuan** [comment]
> > For a RayJob, there will be some [head pods](https://github.com/ray-project/kuberay/blob/b5f14f1a173d32e6e5c5abfecd11d32a4ed780c1/ray-operator/apis/ray/v1/raycluster_types.go#L25) and some worker groups. Worker pods communicate with head pods to run tasks.
> 
> Got it, thanks. Given worker group could have a different pod template, so each worker group can be treated as a component. Right?

Yes, each worker group can be treated as a component.

**Comment by RainbowMango** [comment]
OK. Just added both of them to the issue description.

**Comment by likku123** [comment]
@RainbowMango 

Just a naive question In the present karmada I believe Ray doesn't comes under Karmada built in resource interpreter ? If one has to test Ray job execution via Karmada and Volcano one needs to write up their own customised resource interpreter right.

**Comment by likku123** [comment]
Yes, there are group of people using kuberay cluster for their AI/ML trainings and other group using argoworkflow jobs for their scientific  jobs ( argoworkflow is working as expected after some work arounds).

I think it's better to focus on kuberay as well because of it's popularity and adaptability across the organisation's.

I am just thinking Karmada + volcano combination is what most of the use cases fit into and also widely popular combined solution  to meet majority of the requirements.

How about making the below mentioned ecosystems in volcano make it available as built in Resource interpreter in Karmada?

![Image](https://github.com/user-attachments/assets/a12705fe-4a11-40ff-90e7-2f27af1ad529)



I am just throwing my thoughts here .

**Comment by likku123** [comment]
I agree to your point. Just trying to understand what would be the scale or how one scale or how at Karmada scale on resource type selection. 

My first thumb of rule or preference when trying to explore any tool around k8s is.  Is  the tool backed up by CNCF and the maturity level and also how popular in terms of stars in github and mostly around the ease of availability of documentation.


Would it be possible to include kuberay and Apache airflow as well? 

I see a
Apavhe Airflow adoption is very high in Data Engineers/platform

**Comment by RainbowMango** [comment]
For the Kuberay, i guess you mean the [RayJob](https://github.com/ray-project/kuberay/blob/b81af7c6e01433d51ded5a17edf6bb5a1dc76f88/ray-operator/apis/ray/v1/rayjob_types.go#L71).
This is already included in the issue description.

For the Apache Airflow, could you please share a link?

**Comment by likku123** [comment]
Yes, this is the one [RayJob](https://github.com/ray-project/kuberay/blob/b81af7c6e01433d51ded5a17edf6bb5a1dc76f88/ray-operator/apis/ray/v1/rayjob_types.go#L71).

I am not sure if this is the [one](https://github.com/apache/airflow/tree/main/airflow-core/src/airflow/api) for Apache Airflow.

**Comment by jalonyou** [comment]
> Yes, there are group of people using kuberay cluster for their AI/ML trainings and other group using argoworkflow jobs for their scientific jobs ( argoworkflow is working as expected after some work arounds).
> 
> I think it's better to focus on kuberay as well because of it's popularity and adaptability across the organisation's.
> 
> I am just thinking Karmada + volcano combination is what most of the use cases fit into and also widely popular combined solution to meet majority of the requirements.
> 
> How about making the below mentioned ecosystems in volcano make it available as built in Resource interpreter in Karmada?
> 
> ![Image](https://github.com/user-attachments/assets/a12705fe-4a11-40ff-90e7-2f27af1ad529)
> 
> I am just throwing my thoughts here .

will karmada make above CRDs  built in Resource interprete in V1.14？ is there any clear plan?

**Comment by RainbowMango** [comment]
I don't see any blocker if someone wants to contribute the default Resource Interpreter for CRDs above.
Please let me know which one you want to prioritize, and if you like to help.

Speaking of the plan, I'm hoping to extend the Resource Interpreter framework in release 1.15. Will try to include as many built-in interpreters as we can.
By the way, we just released v1.14 last week.

**Comment by serbrech** [comment]
Any thoughts on leaving this specialized type of integration to something like https://kueue.sigs.k8s.io/docs/concepts/multikueue/ via integration, rather than building it all into Karmada?

**Comment by RainbowMango** [comment]
Hi @serbrech, this feature is not for building a queue management system; it is for more precise resource-aware scheduling for workloads with multiple pod templates.

I think we are open to integrating with any projects from the ecosystem. Speaking of the Kueue, I'm still unsure how to do it. Do you have any idea about it?
